tools/rewriter/GenCallAsm.cpp

```diff
@@ -280,21 +286,378 @@ static std::string sig_string(const CAbiSignature &sig,
+          aw, "Copy stack arguments from the caller's stack to the compartment");
+      // Load addr of top of caller compartment's stack into rax, clobbering r12
+      std::string expr = emit_load_sp_offset(aw, caller_pkey, "r12"s);
+      add_asm_line(aw, "movq %" + expr + ", %rax");
+      // This is effectively a memcpy of size `stack_arg_size` from the caller's
+      // stack to the compartment's
+      for (int i = 0; i < stack_arg_size; i += 8) {
+        // We must take the preserved registers we pushed on the caller's stack
+        // into account (including rbp) when determining the location of the stack
+        // args
+        size_t caller_stack_size =
+            stack_arg_size + ((preserved_registers.size() + 1) * 8);
+        // The index into the caller's stack is backwards since pushq will copy to
+        // the compartment's stack from the highest addresses to the lowest.
+        add_asm_line(aw,
+                     "pushq " + std::to_string(caller_stack_size - i) + "(%rax)");
+      }
+    }
+  } else if (arch == Arch::Aarch64) {
+    // TODO ARM arguments
+    llvm::errs() << "TODO arguments not implemented on ARM\n";
+  }
+}
+
+static void emit_zero_regs(AsmWriter &aw, uint32_t caller_pkey, int reg_arg_count, const std::vector<ParamLocation> &param_locs, Arch arch) {
+  if (arch == Arch::X86) {
+    // Before calling the target function, we only need to zero out registers not
+    // used for arguments if the caller is a protected compartment (i.e. pkey is
+    // not zero).
+    if (caller_pkey != 0) {
+      // Zero out all unused registers. First we save all registers containing args.
+      // These pushes have matching pops before calling the wrapped function so this
+      // stack space is not shown in the diagram above.
+      if (reg_arg_count > 0) {
+        add_comment_line(aw, "Save registers containing arguments");
+        for (const auto &loc : param_locs) {
+          if (!loc.is_stack()) {
+            emit_reg_push(aw, loc);
+          }
+        }
+      }
+
+      // Zero all registers except rsp
+      add_comment_line(aw, "Zero all registers except rsp");
+      // FIXME: If this will use the System V ABI make sure that the %rsp is aligned
+      // before this call
+      add_asm_line(aw, "call __libia2_scrub_registers");
+
+      // Restore used arg regs after zeroing registers
+      if (reg_arg_count > 0) {
+        add_comment_line(aw, "Restore registers containing arguments");
+        for (auto loc = param_locs.rbegin(); loc != param_locs.rend(); loc++) {
+          if (!loc->is_stack()) {
+            emit_reg_pop(aw, *loc);
+          }
+        }
+      }
+    }
+  } else if (arch == Arch::Aarch64) {
+    // TODO ARM reg zero
+    llvm::errs() << "TODO register zeroing not implemented on ARM\n";
+  }
+}
+
+static void emit_set_pkru(AsmWriter &aw, uint32_t target_pkey, WrapperKind kind, Arch arch) {
+  if (arch == Arch::X86) {
+    if (kind == WrapperKind::IndirectCallsite) {
```

Comment:
This branch isn't related to the PKRU so you should pull it out of this function.

Reply:
I moved this out into its own function `emit_fn_ptr`